### PR TITLE
GODRIVER-2443 Make Distinct return a decodable struct

### PR DESCRIPTION
**Pull request by prestonvasquez**
GODRIVER-2443

## Summary

The `Distinct` collection method should return a struct that can be decoded, similar to `FindOne`.

## Background & Motivation

Providing a decodable type will improve user experience. Instead of iterating through an any slice, users can decode same-type data using conventional Go syntax:

```go
res, err := coll.Distinct(ctx, "myField", filter)
if err != nil {
	return err
}

var arr []string
res.Decode(&arr)
```

Alternatively, if the data returned is not same-type a user can iterate it directly as a bson.RawArray type:

```go
res, err := coll.Distinct(ctx, "myField", filter)
if err != nil {
	return err
}

val, err := res.Values()
if err != nil {
	return err
}

for _, rawValue := range val {
	switch rawValue.Type {
		case bson.TypeDouble:
			var f64 float64 
			
			rawValue.Unmarshal(&f64)
			fmt.Println("f64: ", f64)
		case bson.TypeString:
			var str string 
			
			rawValue.Unmarshal(&str)
			fmt.Println("str: ", str)
	}
}
```

### CODE REVIEW

File: mongo/single_result.go [hunk 1]
```diff
@@ -22,11 +22,11 @@ var ErrNoDocuments = errors.New("mongo: no documents in result")
 // SingleResult represents a single document returned from an operation. If the operation resulted in an error, all
 // SingleResult methods will return that error. If the operation did not return any documents, all SingleResult methods
 // will return ErrNoDocuments.
-type SingleResult struct {
+type SingleResult[T bson.Raw | bson.RawArray] struct {
```
Comment (prestonvasquez) [inline mongo/single_result.go:25]: T defines the return type of the `SingleResult.Raw` method.

File: mongo/single_result.go [hunk 2]
```diff
@@ -74,32 +78,37 @@ func (sr *SingleResult) Decode(v interface{}) error {
...
 
-	dec := getDecoder(sr.rdr, sr.bsonOpts, sr.reg)
+	if sr.rdr.Type == bson.TypeEmbeddedDocument {
+		dec := getDecoder(sr.rdr.Value, sr.bsonOpts, sr.reg)
 
-	return dec.Decode(v)
+		return dec.Decode(v)
+	}
+
+	return sr.rdr.Unmarshal(v)
```
Comment (matthewdale) [inline mongo/single_result.go:87]: We need to use the configured `Registry` and `BSONOptions` when unmarshaling the `RawValue` here.

File: mongo/collection.go [hunk 3]
```diff
@@ -1372,30 +1376,25 @@ func (coll *Collection) Distinct(ctx context.Context, fieldName string, filter i
...
-	retArray := make([]interface{}, len(values))
-
-	for i, val := range values {
-		raw := bson.RawValue{Type: val.Type, Value: val.Data}
-		err = raw.Unmarshal(&retArray[i])
-		if err != nil {
-			return nil, err
-		}
+	return &SingleResult[bson.RawArray]{
+		ctx:      ctx,
```
Comment (matthewdale) [inline mongo/collection.go:1390]: This `Context` will never be used in the case of `Distinct` and can be omitted.

File: mongo/single_result.go (hunk 1 above)
Comment (matthewdale) [reply mongo/single_result.go:25]: I'm concerned about the scope of the change required to add this type parameter. One of the challenges with generic types is that they require parameterizing all functions and types that use them, unless only a single type is ever used (see `verifySingleResult`).

An alternate approach that doesn't impact any existing types is to add a new type `DistinctResult` that holds a `SingleResult` and mostly reproduces the API of `SingleResult`, changing the return type of `Raw`. That creates fewer parameterized type changes in existing code and provides the ability to independently extend the APIs of `DistinctResult` and `SingleResult`.

E.g.
```go
type DistinctResult struct {
	sr *SingleResult
}

func (dr *DistinctResult) Decode(v interface{}) error { return dr.sr.Decode(v) }
func (dr *DistinctResult) Err() error                 { return dr.sr.Err() }

func (dr *DistinctResult) Raw() (bson.RawArray, error) {
	r, err := dr.sr.Raw()
	return bson.RawArray(r), err
}
```

File: internal/integration/crud_helpers_test.go [hunk 4]
```diff
@@ -1638,7 +1648,11 @@ func verifyCursorResult(mt *mtest.T, cur *mongo.Cursor, result interface{}) {
 	}
 }
 
-func verifySingleResult(mt *mtest.T, actualResult *mongo.SingleResult, expectedResult interface{}) {
+func verifySingleResult[T bson.Raw | bson.RawArray](
+	mt *mtest.T,
+	actualResult *mongo.SingleResult[T],
+	expectedResult interface{},
+) {
```
Comment (matthewdale) [inline internal/integration/crud_helpers_test.go:1655]: `verifySingleResult` is only ever called with a `bson.Raw`, so the function doesn't need to be parameterized. Then the `bson.Raw` conversion in the call to `compareDocs` can also be removed.

E.g. function signature:
```go
func verifySingleResult(
	mt *mtest.T,
	actualResult *mongo.SingleResult[bson.Raw],
	expectedResult interface{},
) {
```

**Review by matthewdale — CHANGES_REQUESTED** [review]
Overall, the API of `SingleResult` seems to work great for decoding results from `Distinct`. I have concerns about the parameterized type.

**Comment by prestonvasquez** [comment]
@matthewdale  would you mind expanding on this concern: https://github.com/mongodb/mongo-go-driver/pull/1603#discussion_r1580467165 ?  Genericizing `SingleResult` shouldn't require any updates for existing methods that return `SingleResult`. That is, the only user code that would require updating are the usages of `collection.Distinct()`.

File: mongo/single_result.go (hunk 1 above)
Comment (matthewdale) [reply mongo/single_result.go:24]: [Here](https://sourcegraph.com/search?q=context:global+lang:Go++-repo:%5Egithub%5C.com/mongodb/mongo-go-driver%24+repo:has.content%28go.mongodb.org/mongo-driver%29+-file:%28%5E%7C/%29vendor/+mongo.SingleResult&patternType=standard&case=yes&sm=0&groupBy=repo) are some examples of user code that would be impacted by adding a type parameter to `SingleResult`. They seem to fall in a few categories:
* Database client wrappers that return a `*mongo.SingleResult`.
* Interfaces used for mocking the `Collection` API.
* Adding helper functions for using a `*mongo.SingleResult`.

**Comment by matthewdale** [comment]
@prestonvasquez I added some Sourcegraph search results for code that references `mongo.SingleResult` and would be impacted by the change to a comment [here](https://github.com/mongodb/mongo-go-driver/pull/1603#discussion_r1583828594).

File: internal/integration/collection_test.go [hunk 5]
```diff
@@ -874,24 +874,29 @@ func TestCollection(t *testing.T) {
...
-				assert.Nil(mt, err, "Distinct error: %v", err)
-				assert.Equal(mt, tc.expected, res, "expected result %v, got %v", tc.expected, res)
+				res := mt.Coll.Distinct(context.Background(), "x", tc.filter, tc.opts)
+				assert.Nil(mt, res.Err(), "Distinct error: %v", res.Err())
+
+				var got []int32
+				err := res.Decode(&got)
+				assert.NoError(t, err)
+
+				assert.EqualValues(mt, tc.want, got, "expected result %v, got %v", tc.want, res)
```
Comment (qingyang-hu) [inline internal/integration/collection_test.go:899]:
```suggestion
				assert.EqualValues(mt, tc.want, got, "expected result %v, got %v", tc.want, got)
```

File: internal/integration/crud_helpers_test.go [hunk 6]
```diff
@@ -1529,25 +1535,29 @@ func verifyDeleteResult(mt *mtest.T, res *mongo.DeleteResult, result interface{}
...
-func verifyDistinctResult(mt *mtest.T, actualResult []interface{}, expectedResult interface{}) {
+func verifyDistinctResult(
+	mt *mtest.T,
+	got bson.RawArray,
+	want interface{},
+) {
 	mt.Helper()
 
-	if expectedResult == nil {
+	if got == nil {
```
Comment (qingyang-hu) [inline internal/integration/crud_helpers_test.go:1545]: Do you need to sanity check `want` as well?

File: mongo/single_result.go (hunk 1 above)
Comment (prestonvasquez) [reply mongo/single_result.go:24]: I understand your point. It's worth noting that most of these examples will require updates due to GODRIVER-2696. However, I believe it's reasonable to minimize the impact where possible

File: mongo/single_result.go (hunk 2 above)
Comment (prestonvasquez) [reply mongo/single_result.go]: It makes sense to decode distinct data with a registry, but the BSONOptions are specific to `bson.Decoder` which expects the value to decode into to be a map / struct. E.g., the following would throw a runtime error:

```go
var i32s []int32
distincResult.Decode(i32s)
```

Comment (matthewdale) [reply mongo/single_result.go]: You're correct, there's currently no way to do that. However, when combined with the changes in https://github.com/mongodb/mongo-go-driver/pull/1584, the following code using a `bson.Decoder` to decode a value works:
```go
func (dr *DistinctResult) Decode(v any) error {
	doc := bsoncore.NewDocumentBuilder().
		AppendValue("Arr", bsoncore.Value{
			Type: bsoncore.TypeArray,
			Data: dr.arr,
		}).Build()

	dec := getDecoder(doc, dr.opts, dr.reg)

	xs := struct {
		Arr any
	}{
		Arr: v,
	}
	return dec.Decode(&xs)
}
```
I created [GODRIVER-3205](https://jira.mongodb.org/browse/GODRIVER-3205) and added it to the Go Driver 2.0 epic to track that improvement, so we can consider that is resolved for now.

**Review by matthewdale — DISMISSED** [review]
Looks good! 👍

**Review by qingyang-hu — DISMISSED** [review]
👍

**Review by matthewdale — APPROVED** [review]

**Review by qingyang-hu — APPROVED** [review]

**Pull request merged**